REGRESSION(287869@main?): [macOS Debug wk2] ASSERTION FAILED: frame->presentationTime() >= m_lastMuxedSampleStartTime in WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame()

jyavenard opened the pull request:
#### f75e1eb03f2587e613a564ac94c78089b07beb95
<pre>
REGRESSION(287869@main?): [macOS Debug wk2] ASSERTION FAILED: frame-&gt;presentationTime() &gt;= m_lastMuxedSampleStartTime in WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame()
<a href="https://bugs.webkit.org/show_bug.cgi?id=284853">https://bugs.webkit.org/show_bug.cgi?id=284853</a>
<a href="https://rdar.apple.com/141646476">rdar://141646476</a>

Reviewed by Youenn Fablet.

This is a tentative fix involving four parts:
- We remove the assertion as it was mostly in place for development purposes; instead we log the errors.
  The assertion would normally occur if the incoming data was going to cause an invalid webm (non monotonically increasing timestamps).
- Add assertion that the compressed frames timestamp are increasing monotonically as it&apos;s one explanation that
  would trigger the earlier assertion with http/wpt/mediarecorder/MediaRecorder-video-bitrate.html
- Calculate the incoming video frame time at the time it is received to reduce the likelihood of frames
  having the same timestamp.
- Drop incoming video frames as soon as the pause() has been called as it could have caused
  an invalid calculation of the video frame time due to inconsistency between m_currentVideoDuration and m_resumeWallTime

Covered by existing tests, re-enable tests skipped in 287971@main.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::pause):
(WebCore::MediaRecorderPrivateEncoder::resume):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame):
(WebCore::MediaRecorderPrivateEncoder::currentTime const):
(WebCore::MediaRecorderPrivateEncoder::currentEndTime const):
(WebCore::MediaRecorderPrivateEncoder::nextVideoFrameTime): Deleted. Perform the calculation in appendVideoFrame instead.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: <a href="https://commits.webkit.org/288583@main">https://commits.webkit.org/288583@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b0dda42ddaa146125f43660ef22405b2e9e4039

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3426 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88881 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34816 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85893 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11391 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65186 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23019 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86854 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76151 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45475 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2521 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30367 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33865 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31122 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90258 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7998 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73629 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11297 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71979 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72851 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18017 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17123 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2397 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11025 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16497 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10873 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14348 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12645 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
```diff
@@ -175,9 +175,9 @@ class MediaRecorderPrivateEncoder final : public ThreadSafeRefCountedAndCanMakeT
     MediaTime m_lastVideoKeyframeTime WTF_GUARDED_BY_CAPABILITY(queueSingleton());
     MediaTime m_lastReceivedCompressedVideoFrame WTF_GUARDED_BY_CAPABILITY(queueSingleton()) { MediaTime::negativeInfiniteTime() };
-    MediaTime m_currentVideoDuration WTF_GUARDED_BY_CAPABILITY(queueSingleton());
+    std::atomic<int64_t> m_currentVideoDurationUs { 0 };
```
```suggestion
    std::atomic<uint64_t> m_currentVideoDurationUs { 0 };
```

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -463,7 +465,13 @@ void MediaRecorderPrivateEncoder::appendVideoFrame(const MediaTime& audioTime, R
 
-    auto sampleTime = nextVideoFrameTime(audioTime);
+    auto sampleTime = time;
+    // We take the time before m_firstVideoFrameAudioTime is set so that the first frame will always apepar to have a timestamp of 0 but with a longer duration.
```
```suggestion
    // We take the time before m_firstVideoFrameAudioTime is set so that the first frame will always appear to have a timestamp of 0 but with a longer duration.
```

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -411,20 +414,19 @@ void MediaRecorderPrivateEncoder::audioSamplesAvailable(const MediaTime& time, s
 }
 
-void MediaRecorderPrivateEncoder::appendVideoFrame(const MediaTime& audioTime, Ref<VideoFrame>&& frame)
+void MediaRecorderPrivateEncoder::appendVideoFrame(const MediaTime& time, Ref<VideoFrame>&& frame)
```
```suggestion
void MediaRecorderPrivateEncoder::appendVideoFrame(MediaTime sampleTime, Ref<VideoFrame>&& frame)
```
This will remove the need to redeclare `sampleTime`.

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -411,20 +414,19 @@ void MediaRecorderPrivateEncoder::audioSamplesAvailable(const MediaTime& time, s
 void MediaRecorderPrivateEncoder::appendVideoFrame(VideoFrame& frame)
 {
-    if (m_isStopped)
+    if (m_isStopped || m_isPaused)
```
We somehow loose a few frames when resuming.
In theory, we should probably check for m_isPaused in the lambda below.
And we could capture `now()` to set m_resumeWallTime if needed.

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -227,15 +229,16 @@ void MediaRecorderPrivateEncoder::resume()
     assertIsMainThread();
 
+    if (!hasAudio())
+        m_resumeWallTime = MonotonicTime::now();
```
It is a bit odd to set `m_resumeWallTime` here and in `appendVideoFrame` as well.
Should we add a `!m_resumeWallTime.load()` check as well?

jyavenard replied:
This is from the original MediaRecorder implementation (that you wrote :) ). It is only used to determine the duration of the video. In `appendVideoFrame` it deals on setting it for the first time ever, which will be used to set the time of the first frame.

When we pause, `now()-m_resumeWallTime` becomes the duration of the video and `m_resumeWallTime` is then set upon the next resume so that we can determine the duration of the new video segment.

I could probably makes this code a bit easier to follow.

jyavenard replied:
I've reworked things a bit so that it's now all done in the workQueue and renamed the member to `m_currentVideoSegmentStartTime` .

jyavenard committed:
REGRESSION(287869@main?): [macOS Debug wk2] ASSERTION FAILED: frame->presentationTime() >= m_lastMuxedSampleStartTime in WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame()
https://bugs.webkit.org/show_bug.cgi?id=284853
rdar://141646476

Reviewed by Youenn Fablet.

This is a tentative fix involving four parts:
- We remove the assertion as it was mostly in place for development purposes; instead we log the errors.
  The assertion would normally occur if the incoming data was going to cause an invalid webm (non monotonically increasing timestamps).
- Add assertion that the compressed frames timestamp are increasing monotonically as it's one explanation that
  would trigger the earlier assertion with http/wpt/mediarecorder/MediaRecorder-video-bitrate.html
- Calculate the incoming video frame time at the time it is received to reduce the likelihood of frames
  having the same timestamp.
- Drop incoming video frames as soon as the pause() has been called as it could have caused
  an invalid calculation of the video frame time due to inconsistency between m_currentVideoDuration and m_resumeWallTime

Covered by existing tests, re-enable tests skipped in 287971@main.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::pause):
(WebCore::MediaRecorderPrivateEncoder::resume):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame):
(WebCore::MediaRecorderPrivateEncoder::currentTime const):
(WebCore::MediaRecorderPrivateEncoder::currentEndTime const):
(WebCore::MediaRecorderPrivateEncoder::nextVideoFrameTime): Deleted. Perform the calculation in appendVideoFrame instead.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: https://commits.webkit.org/288583@main
